cronolog.1
.TH cronolog 1 "December 1996"
.\" $Id: cronolog.1,v 1.2 1996/12/15 15:34:44 andrew Exp $
.\" Copyright (c) 1996 Andrew Ford.  All rights reserved.
.\" The file LICENSE specifies the terms and conditions for redistribution.
.SH NAME
cronolog \- write log messages to log files named according to a template
.SH SYNOPSIS
.B cronolog
.I template
.SH DESCRIPTION
.B cronolog
is a simple program that reads log messages from its input and writes
them to a set of output files, the names of which are constructed
using
.B template
and the current date and time.  The template uses the same format
specifiers as the Unix
.BR date (1)
command (which are the same as the standard C strftime library
function).  
.PP
Before writing a message
.B cronolog
checks the time to see whether the current log file is still valid and
if not it closes the current file, expands the template using the
current date and time to generate a new file name, opens the new file
(creating missing directories on the path of the new log file as
needed unless the program is compiled with -DDONT_CREATE_SUBDIRS) and
calculates the time at which the new file will become invalid.
.PP
.B cronolog
is intended to be used in conjunction with a Web server, such as
Apache to split the access log into daily or monthly logs.  For
example the Apache configuration directives:
.LP
        TransferLog "|/www/sbin/cronolog /www/logs/%Y/%m/%d/access.log"
        ErrorLog    "|/www/sbin/cronolog /www/logs/%Y/%m/%d/errors.log"
.LP
would instruct Apache to pipe its access and error log messages into
separate copies of cronolog, which would create new log files each day
in a directory hierarchy structured by date, i.e. on 31 December 1996
messages would be written to
.LP
        /www/logs/1996/12/31/access.log
        /www/logs/1996/12/31/errors.log
.P
after midnight the files
.LP
        /www/logs/1997/01/01/access.log
        /www/logs/1997/01/01/errors.log
.P
would be used, with the directories 1997, 1997/01 and 1997/01/01 being
created if they did not already exist.
.SS Template format
Each character in the template represents a character in the expanded
filename, except for date and time format specifiers, which are
replaced by their expansion.  Format specifiers consist of a `%'
followed by one of the following characters:
.IP %
a literal % character
.IP n
a new-line character
.IP t
a horizontal tab character
.PP
Time fields:
.IP H
hour (00..23)
.IP I
hour (01..12)
.IP p
the locale's AM or PM indicator
.IP M
minute (00..59)
.IP S
second (00..61, which allows for leap seconds)
.IP X
the locale's time representation (e.g.: "15:12:47")
.IP Z
time zone (e.g. GMT), or nothing if the time zone cannot be determined
.PP
Date fields:
.IP a
the locale's abbreviated weekday name (e.g.: Sun..Sat)
.IP A
the locale's full weekday name (e.g.: Sunday .. Saturday)
.IP b
the locale's abbreviated month name (e.g.: Jan .. Dec)
.IP B
the locale's full month name, (e.g.: January .. December)
.IP c
the locale's date and time (e.g.: "Sun Dec 15 14:12:47 GMT 1996")
.IP d
day of month (01 .. 31)
.IP j
day of year (001 .. 366)
.IP m
month (01 .. 12)
.IP U
week of the year with Sunday as first day of week (00..53, where week
1 is the week containing the first Sunday of the year)
.IP W
week of the year with Monday as first day of week (00..53, where week
1 is the week containing the first Monday of the year)
.IP w
day of week (0 .. 6, where 0 corresponds to Sunday)
.IP x
locale's date representation (e.g. today in Britain: "15/12/96")
.IP y
year without the century (00 .. 99)
.IP Y
year with the century (1970 .. 2038)
.P
Other specifiers may be available depending on the C library's
implementation of the
.B strftime
function.
.P
.SH SEE ALSO
.BR apache (1m)
.BR date (1)
.BR strftime (3)
.BR environ (5)
.PP
More information and the latest version of 
.B cronolog 
can be obtained from
.LP
        http://www.nhbs.co.uk/aford/resources/apache/cronolog/
.LP
If you have any suggestions, bug reports, fixes, or enhancements,
please mail them to the author.
.SS More about Apache
Documentation for the Apache http server is available from
.LP
        http://www.apache.org
.SH FUTURE DIRECTIONS
The functionality of
.B cronolog
could be built into Apache, thus saving the overhead of having a
process per log stream and that of transferring data from the server
process to the cronolog process.  The main complication is handling
the case where multiple log streams are written to the same file
(template), for example where different virtual servers write to the
same set of log files.
.SH AUTHOR
Andrew Ford <andrew@icarus.demon.co.uk> (http://www.nhbs.co.uk/aford).
.LP
.B cronolog
is based on a program called
.B rotatelogs
by Ben Laurie, which is packaged with the Apache web server.


NEWS
1.6.1 (20-Dec-1999)

* Cronolog submitted to Apache Software Foundation


1.6 (16-Dec-1999)

* Fixed bug in cronolog whereby if it was started shortly before the end of 
  a period it would not rotate until the end of the next period.
  (Thanks to  Charles Thayer <thayer@mediabridge.com> for identifying the bug 
  and supplying the fix.)


1.5b9 (04-Jun-1998)

* Fixed typo in cronolog.c and cronotest.c  [[ if (start_time); { ... } ]]
  (gcc -Wall doesn't even flag it!)

* changed definition of multi-line string constants to use proper ANSI string 
  concatenation 


1.5b8 (08-Mar-1998)

* Log files only created after a message has been read (you shouldn't
  get empty files)

* time() called only once per message 

* Extra debugging code (see README)

* Updated documentation


Major changes in release 1.5 (11-Apr-1997):

* Fixed problem with change to daylight saving time (summer time).

* Restructured to conform to GNU standards (now uses GNU automake and autoconf).

Major changes in release 1.4 (20-Dec-1996):

* Optimization of create_subdirs: don't bother stat'ing directories
  that form part of a common prefix with the last filename checked.


Major changes in release 1.3 (16-Dec-1996):

* Fixed portability bugs. 


Major changes in release 1.2 (15-Dec-1996):

* Creates missing directories as needed.

* First version announced on comp.infosys.www.servers.unix (renamed to cronolog).

   
Major changes in release 1.1 (09-Dec-1996):

* Fixed problem with log files being created with the wrong dates.

  (When a new log file is due to be created, the time used for filling
  out the template should be the start of the current period and not
  the time at which a new log file is due.  This was observed with daily
  logs where there were no transfers for two days; the next log file
  generated had the date of the first of the days with no transfers.)

* Also added more comments and fixed handling of week numbers where
  weeks start on a Monday.


Major changes in release 1.0 (06-Dec-1996):

* Initial version sent to Apache developers' mailing list 
  (at that time cronolog was called strftimelog)


src/cronolog.c
/* ====================================================================
 * Copyright (c) 1995-1999 The Apache Group.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer. 
 *
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in
 *    the documentation and/or other materials provided with the
 *    distribution.
 *
 * 3. All advertising materials mentioning features or use of this
 *    software must display the following acknowledgment:
 *    "This product includes software developed by the Apache Group
 *    for use in the Apache HTTP server project (http://www.apache.org/)."
 *
 * 4. The names "Apache Server" and "Apache Group" must not be used to
 *    endorse or promote products derived from this software without
 *    prior written permission. For written permission, please contact
 *    apache@apache.org.
 *
 * 5. Products derived from this software may not be called "Apache"
 *    nor may "Apache" appear in their names without prior written
 *    permission of the Apache Group.
 *
 * 6. Redistributions of any form whatsoever must retain the following
 *    acknowledgment:
 *    "This product includes software developed by the Apache Group
 *    for use in the Apache HTTP server project (http://www.apache.org/)."
 *
 * THIS SOFTWARE IS PROVIDED BY THE APACHE GROUP ``AS IS'' AND ANY
 * EXPRESSED OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE APACHE GROUP OR
 * ITS CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 * ====================================================================
 *
 * This software consists of voluntary contributions made by many
 * individuals on behalf of the Apache Group and was originally based
 * on public domain software written at the National Center for
 * Supercomputing Applications, University of Illinois, Urbana-Champaign.
 * For more information on the Apache Group and the Apache HTTP server
 * project, please see <http://www.apache.org/>.
 *
 */
/*
 * cronolog -- simple Apache log rotation program
 *
 * Copyright (c) 1996-1999 by Ford & Mason Ltd
 *
 * This software was submitted by Ford & Mason Ltd to the Apache
 * Software Foundation in December 1999.  Future revisions and
 * derivatives of this source code must acknowledge Ford & Mason Ltd
 * as the original contributor of this module.  All other licensing
 * and usage conditions are those of the Apache Software Foundation.
 *
 * Originally written by Andrew Ford <A.Ford@ford-mason.co.uk>
 *
 * cronolog is loosly based on the rotatelogs program, which is part of the
 * Apache package written by Ben Laurie <ben@algroup.co.uk>
 *
 * The argument to this program is the log file name template as an
 * strftime format string.  For example to generate new error and
 * access logs each day stored in subdirectories by year, month and day add
 * the following lines to the httpd.conf:
 *
 *	TransferLog "|/www/etc/cronolog /www/logs/%Y/%m/%d/access.log"
 *	ErrorLog    "|/www/etc/cronolog /www/logs/%Y/%m/%d/error.log"
 *
 * The option "-x file" specifies that debugging messages should be
 * written to "file" (e.g. /dev/console) or to stderr if "file" is "-".
 */

#include "cronoutils.h"
#include "getopt.h"


/* Forward function declaration */

int	new_log_file(const char *, const char *, mode_t, const char *,
		     PERIODICITY, int, int, char *, size_t, time_t, time_t *);


/* Definition of version and usage messages */

#ifndef _WIN32
#define VERSION_MSG   	PACKAGE " version " VERSION "\n"
#else
#define VERSION_MSG      "cronolog version 0.1\n"
#endif


#define USAGE_MSG 	"usage: %s [OPTIONS] logfile-spec\n" \
			"\n" \
			"   -H NAME,   --hardlink=NAME maintain a hard link from NAME to current log\n" \
			"   -S NAME,   --symlink=NAME  maintain a symbolic link from NAME to current log\n" \
			"   -P NAME,   --prev-symlink=NAME  maintain a symbolic link from NAME to previous log\n" \
			"   -l NAME,   --link=NAME     same as -S/--symlink\n" \
			"   -h,        --help          print this help, then exit\n" \
			"   -p PERIOD, --period=PERIOD set the rotation period explicitly\n" \
			"   -d DELAY,  --delay=DELAY   set the rotation period delay\n" \
			"   -o,        --once-only     create single output log from template (not rotated)\n" \
			"   -x FILE,   --debug=FILE    write debug messages to FILE\n" \
			"                              ( or to standard error if FILE is \"-\")\n" \
			"   -a,        --american         American date formats\n" \
			"   -e,        --european      European date formats (default)\n" \
			"   -s TIME,   --start-time=TIME   starting time\n" \
			"   -z TZ,     --time-zone=TZ  use TZ for timezone\n" \
			"   -V,        --version       print version number, then exit\n"


/* Definition of the short and long program options */

char          *short_options = "ad:eop:s:z:H:P:S:l:hVx:";

#ifndef _WIN32
struct option long_options[] =
{
    { "american",	no_argument,		NULL, 'a' },
    { "european",	no_argument,		NULL, 'e' },
    { "start-time", 	required_argument,	NULL, 's' },
    { "time-zone",  	required_argument,	NULL, 'z' },
    { "hardlink",  	required_argument, 	NULL, 'H' },
    { "symlink",   	required_argument, 	NULL, 'S' },
    { "prev-symlink",  	required_argument, 	NULL, 'P' },
    { "link",      	required_argument, 	NULL, 'l' },
    { "period",		required_argument,	NULL, 'p' },
    { "delay",		required_argument,	NULL, 'd' },
    { "once-only", 	no_argument,       	NULL, 'o' },
    { "help",      	no_argument,       	NULL, 'h' },
    { "version",   	no_argument,       	NULL, 'V' }
};
#endif

/* Main function.
 */
int
main(int argc, char **argv)
{
    PERIODICITY	periodicity = UNKNOWN;
    PERIODICITY	period_delay_units = UNKNOWN;
    int		period_multiple = 1;
    int		period_delay  = 0;
    int		use_american_date_formats = 0;
    char 	read_buf[BUFSIZE];
    char 	tzbuf[BUFSIZE];
    char	filename[MAX_PATH];
    char	*start_time = NULL;
    char	*template;
    char	*linkname = NULL;
    char	*prevlinkname = NULL;
    mode_t	linktype = 0;
    int 	n_bytes_read;
    int		ch;
    time_t	time_now;
    time_t	time_offset = 0;
    time_t	next_period = 0;
    int 	log_fd = -1;

#ifndef _WIN32
    while ((ch = getopt_long(argc, argv, short_options, long_options, NULL)) != EOF)
#else
    while ((ch = getopt(argc, argv, short_options)) != EOF)
#endif        
    {
	switch (ch)
	{
	case 'a':
	    use_american_date_formats = 1;
	    break;
	    
	case 'e':
	    use_american_date_formats = 0;
	    break;
	    
	case 's':
	    start_time = optarg;
	    break;

	case 'z':
	    sprintf(tzbuf, "TZ=%s", optarg);
	    putenv(tzbuf);
	    break;

	case 'H':
	    linkname = optarg;
	    linktype = S_IFREG;
	    break;

	case 'l':
	case 'S':
	    linkname = optarg;
#ifndef _WIN32
	    linktype = S_IFLNK;
#endif        
	    break;
	    
	case 'P':
	    if (linkname == NULL)
	    {
		fprintf(stderr, "A current log symlink is needed to mantain a symlink to the previous log\n", argv[0]);
		exit(1);
	    }
	    prevlinkname = optarg;
	    break;
	    

	case 'd':
	    period_delay_units = parse_timespec(optarg, &period_delay);
	    break;

	case 'p':
	    periodicity = parse_timespec(optarg, &period_multiple);
	    if (   (periodicity == INVALID_PERIOD)
		|| (periodicity == PER_SECOND) && (60 % period_multiple)
		|| (periodicity == PER_MINUTE) && (60 % period_multiple)
		|| (periodicity == HOURLY)     && (24 % period_multiple)
		|| (periodicity == DAILY)      && (period_multiple > 365)
		|| (periodicity == WEEKLY)     && (period_multiple > 52)
		|| (periodicity == MONTHLY)    && (12 % period_multiple)) {
		fprintf(stderr, "%s: invalid explicit period specification (%s)\n", argv[0], start_time);
		exit(1);
	    }		
	    break;
	    
	case 'o':
	    periodicity = ONCE_ONLY;
	    break;
	    
	case 'x':
	    if (strcmp(optarg, "-") == 0)
	    {
		debug_file = stderr;
	    }
	    else
	    {
		debug_file = fopen(optarg, "a+");
	    }
	    break;
	    
	case 'V':
	    fprintf(stderr, VERSION_MSG);
	    exit(0);
	    
	case 'h':
	case '?':
	    fprintf(stderr, USAGE_MSG, argv[0]);
	    exit(1);
	}
    }

    if ((argc - optind) != 1)
    {
	fprintf(stderr, USAGE_MSG, argv[0]);
	exit(1);
    }

    DEBUG((VERSION_MSG "\n"));

    if (start_time)
    {
	time_now = parse_time(start_time, use_american_date_formats);
	if (time_now == -1)
	{
	    fprintf(stderr, "%s: invalid start time (%s)\n", argv[0], start_time);
	    exit(1);
	}
	time_offset = time_now - time(NULL);
	DEBUG(("Using offset of %d seconds from real time\n", time_offset));
    }

    /* The template should be the only argument.
     * Unless the -o option was specified, determine the periodicity.
     */
    
    template = argv[optind];
    if (periodicity == UNKNOWN)
    {
	periodicity = determine_periodicity(template);
    }


    DEBUG(("periodicity = %d %s\n", period_multiple, periods[periodicity]));

    if (period_delay) {
	if (   (period_delay_units > periodicity)
	    || (   period_delay_units == periodicity
		&& abs(period_delay)  >= period_multiple)) {
	    fprintf(stderr, "%s: period delay cannot be larger than the rollover period\n", argv[0], start_time);
	    exit(1);
	}		
	period_delay *= period_seconds[period_delay_units];
    }

    DEBUG(("Rotation period is per %d %s\n", period_multiple, periods[periodicity]));


    /* Loop, waiting for data on standard input */

    for (;;)
    {
	/* Read a buffer's worth of log file data, exiting on errors
	 * or end of file.
	 */
	n_bytes_read = read(0, read_buf, sizeof read_buf);
	if (n_bytes_read == 0)
	{
	    exit(3);
	}
	if (errno == EINTR)
	{
	    continue;
	}
	else if (n_bytes_read < 0)
	{
	    exit(4);
	}

	time_now = time(NULL) + time_offset;
	
	/* If the current period has finished and there is a log file
	 * open, close the log file
	 */
	if ((time_now >= next_period) && (log_fd >= 0))
	{
	    close(log_fd);
	    log_fd = -1;
	}
	
	/* If there is no log file open then open a new one.
	 */
	if (log_fd < 0)
	{
	    log_fd = new_log_file(template, linkname, linktype, prevlinkname,
				  periodicity, period_multiple, period_delay,
				  filename, sizeof (filename), time_now, &next_period);
	}

	DEBUG(("%s (%d): wrote message; next period starts at %s (%d) in %d secs\n",
	       timestamp(time_now), time_now, 
	       timestamp(next_period), next_period,
	       next_period - time_now));

	/* Write out the log data to the current log file.
	 */
	if (write(log_fd, read_buf, n_bytes_read) != n_bytes_read)
	{
	    perror(filename);
	    exit(5);
	}
    }

    /* NOTREACHED */
    return 1;
}
